torch/fx/experimental/symbolic_shapes.py

```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
 DimConstraint = Union[StrictMinMaxConstraint, RelaxedUnspecConstraint, None]
 
 @dataclass(frozen=True)
-class EqualityConstraint(Constraint):
+class EqualityConstraints(Constraint):
     """
-    Given pairs of sources corresponding to pairs of dynamic dimensions that
-    are specified equal, represent them in a union-find data structure so that
-    we can efficiently check whether two such sources are transitively equal.
+    Represent and decide various kinds of equality constraints between input sources.
+
+    A "source pair" is a pair of input sources for dynamic dimensions that
+    are specified equal. We represent `source_pairs` in a union-find forest
+    so that we can efficiently check whether two such sources are transitively equal.
+
+    A "derived equality" relates an input source to an expression over a root.
+    The root can be another input source, corresponding to some dynamic dimension,
+    or a phantom symbol that does not directly represent any dynamic dimension. We
+    represent `derived_equalities` involving input sources in a transitively-closed map
+    so that we can efficiently check whether an input source is transitively equal to
+    a given expression over another input source.
+    (NOTE: In contrast, it is easy to decide whether an input source is transitively equal
+    to a given expression over a phantom symbol; such expressions are already in canonical
+    form and so the problem reduces to symbolic expression equality.)
     """
     source_pairs: List[Tuple[Source, Source]]
+    derived_equalities: List[Tuple[Source, Union[Source, sympy.Symbol], Callable]]
+    phantom_symbols: Dict[str, sympy.Symbol]
```

Comment:
What exactly is the string here?

Reply:
It is the name of a phantom symbol. When building `EqualityConstraints`, the `phantom_symbols` field is populated by `dynamic_shapes.process_equalities`, where we use the key to point to sympy symbols created with the corresponding names.

When using `EqualityConstraints` in symbolic_shapes, the key doesn't really matter. So we could technically just pass in the values of the dict here and change the type to `List[sympy.Symbol]`.